File: src/fumaresult2table/leadSNP2gene.py
import sys
import os
current = os.path.dirname(os.path.realpath(__file__))
parent = os.path.dirname(current)
sys.path.append(parent)

from util import *
from packages import *

def leadSNP2GeneMapping(result_dir, outd, max_dis = 10, verbose=False):
	"""
	max_dis: in kb
	"""
	max_dis = max_dis * 1000

	snp2gene_file = os.path.join(result_dir, "genes.txt")
	assert FileExists(snp2gene_file), "Place `genes.txt` in the result directory."

	logs = []
	logs.append("Running lead SNP to gene mapping to table...")

	df = pd.read_csv(snp2gene_file, sep="\t", index_col=False)

	df = df[['ensg', 'symbol', 'type', 'chr', 'start', 'end', 'ciMap', 'posMapSNPs', 'posMapMaxCADD', 'eqtlMapSNPs', 'eqtlMapminP', 'eqtlMapminQ', 'eqtlMapts', 'ciMapts']]

	df.rename(columns={'ensg':'Ensembl ID', 
					'symbol':'Gene', 
					'type': 'Gene type', 
					'chr': 'Chr', 
					'start': 'Start', 
					'end': 'End', 
					'ciMap': 'ciMap', 
					'posMapSNPs': 'N posMap', 
					'posMapMaxCADD': 'Max CADD posMap', 
					'eqtlMapSNPs': 'N eqtlMap', 
					'eqtlMapminP': 'min P eqtlMap', 
					'eqtlMapts': 'Tissue eqtlMap', 
					'ciMapts': 'Tissue ciMap'}, inplace=True)

	df['posMap'] = df['N posMap'].apply(lambda x: 'Yes' if x > 0 else 'No')
	df['eqtlMap'] = df['N eqtlMap'].apply(lambda x: 'Yes' if x > 0 else 'No')

	## Match which SNP the gene was mapped to
	# positional mapping: `N posMap`
	snps_file = os.path.join(result_dir, "snps.txt")
	df_snps = pd.read_csv(snps_file, sep="\t", index_col=False)
	df_snps = df_snps[['uniqID', 'rsID', 'chr', 'pos', 'posMapFilt']]

	def match_snps4posmap(df_snps, chr, start, end, max_dis):
		df_snps_subset = df_snps[df_snps['chr'] == chr]
		df_snps_subset = df_snps_subset[df_snps_subset['pos'].between(start - max_dis, end + max_dis)]
		if df_snps_subset.empty:
			return '-'
		snps = df_snps_subset[df_snps_subset['posMapFilt'] == 1]['rsID'].tolist()
		return ', '.join(snps)

	df['SNPs posMap'] = df.apply(lambda row: match_snps4posmap(df_snps, row['Chr'], row['Start'], row['End'], max_dis), axis=1)
	
	# eqtl mapping: 
	eqtl_file = os.path.join(result_dir, "eqtl.txt")
	df_eqtl = pd.read_csv(eqtl_file, sep="\t", index_col=False)
	df_eqtl = df_eqtl[['uniqID', 'symbol', 'eqtlMapFilt']]

	def match_snps4eqtlmap(df_eqtl, df_snps, gene):
		df_eqtl_subset = df_eqtl[(df_eqtl['symbol'] == gene) & (df_eqtl['eqtlMapFilt'] == 1)]
		if df_eqtl_subset.empty:
			return '-'
		snps = df_eqtl_subset['uniqID'].tolist()
		df_snps = df_snps[df_snps['uniqID'].isin(snps)][['uniqID', 'rsID']]
		df_eqtl_subset = df_eqtl_subset.merge(df_snps, how="left", on='uniqID')
		df_eqtl_subset.fillna({'rsID':'-9'}, inplace=True)
		df_eqtl_subset['rsID'] = df_eqtl_subset.apply(lambda row: row['uniqID'] if row['rsID'] == '-9' else row['rsID'], axis=1)
		return ', '.join(df_eqtl_subset['rsID'].tolist())

	df['SNPs eqtlMap'] = df.apply(lambda row: match_snps4eqtlmap(df_eqtl, df_snps, row['Gene']), axis=1)

	# chromatin interaction mapping: 
	ci_file = os.path.join(result_dir, 'ci.txt')
	df_ci = pd.read_csv(ci_file, sep="\t", index_col=False)
	df_ci = df_ci[['SNPs', 'genes', 'ciMapFilt']]
	df_ci.fillna({'genes':'NA'}, inplace=True)

	def match_snps4cimap(df_ci, gene_id):
		df_ci_subset = df_ci[(df_ci['genes'].str.contains(gene_id)) & (df_ci['ciMapFilt'] == 1)]
		if df_ci_subset.empty:
			return '-'
		snps = [v.split(sep=":") for v in df_ci_subset['SNPs'].tolist()]
		snps = list(set([item for sublist in snps for item in sublist]))
		return ', '.join(snps)

	df['SNPs ciMap'] = df.apply(lambda row: match_snps4cimap(df_ci, row['Ensembl ID']), axis=1)

	## Re-order columns
	df = df[['Ensembl ID', 'Gene', 'Gene type', 'Chr', 'Start', 'End', 'posMap', 'eqtlMap', 'ciMap',
		'N posMap', 'Max CADD posMap', 'N eqtlMap', 'min P eqtlMap', 'eqtlMapminQ', 'Tissue eqtlMap',
		'Tissue ciMap', 'SNPs posMap', 'SNPs eqtlMap', 'SNPs ciMap']]

	logs.append("Number of mapped gene by positional mapping: {:,}".format(len(df[df['posMap'] == 'Yes'])))
	logs.append("Number of mapped gene by eQTL mapping: {:,}".format(len(df[df['eqtlMap'] == 'Yes'])))
	logs.append("Number of mapped gene by chromatin interaction mapping: {:,}".format(len(df[df['ciMap'] == 'Yes'])))

	df.fillna("-", inplace=True)
	df['Tissue eqtlMap'] = df['Tissue eqtlMap'].apply(lambda x: x.replace(":", ", "))
	df['Tissue eqtlMap'] = df['Tissue eqtlMap'].apply(lambda x: x.replace("_", " "))
	df['Tissue ciMap'] = df['Tissue ciMap'].apply(lambda x: x.replace(":", ", "))
	df['Tissue ciMap'] = df['Tissue ciMap'].apply(lambda x: x.replace("_", " "))


	df.to_csv(os.path.join(outd, "leadSNP_to_gene_mapping_table.csv"), sep=",", index=False, quoting=csv.QUOTE_ALL)
	
	logs.append("Number of genes identified in all 3 strategies: {:,}".format(len(df[(df['posMap'] == 'Yes') & (df['eqtlMap'] == 'Yes') & (df['ciMap'] == 'Yes')])))
	logs.append("")

	if verbose:
		[print(l) for l in logs]
	
	with open(os.path.join(outd, "leadSNP_to_gene_mapping_table.log"), 'w') as f:
		f.writelines([l+"\n" for l in logs])

File: src/fumaresult2table/leadSNP2Gene.R
options(scipen=999, error=traceback) # No scientific number
library(data.table)
library(dplyr)
library(tidyr)
library(argparse)

### Pre-defined function
remove_trailing_slash <- function(input_string) {
    # Replace a forward slash at the end of the string with an empty string
    return(sub("/$", "", input_string))
}

delim_map <- c("comma" = ",", "tab" = "\t", "whitespace" = " ")
delim_extension_map <- c("comma" = ".csv", "tab" = ".tsv", "whitespace" = ".txt")

### Parse command line arguments
parser <- ArgumentParser()
parser$add_argument("--dir-fuma", dest="dir_fuma", type = "character", required = TRUE,
                    help="Directory with FUMA results.")

parser$add_argument("--posMap-max-distance", dest = "posMap_max_distance", type = "integer", required = FALSE, default = 10,
                    help="Maximum distance used for position mapping.")
parser$add_argument("--thread", dest="n_thread", type = "integer", required = FALSE, default = 1, 
                    help="Number of threads to use.")

parser$add_argument("--remove-gene-placeholder", dest="remove_gene_placeholder", 
                    required = FALSE, action = "store_true", default = FALSE,
                    help="Remove gene placeholder.")

parser$add_argument("--dir-out", dest = "dir_out", type = "character", required = FALSE, default = ".",
                    help="Output directory. Default = current working directory.")
parser$add_argument("--name-out", dest = "name_out", type = "character", required = FALSE, 
                    default = "FUMA.gene_mapping.combined",
                    help="Output file prefix. Default = 'FUMA.gene_mapping.combined'")
parser$add_argument("--delim-out", dest = "delim_out", type = "character", required = FALSE, default = "comma",
                    help="Output file delimiter. Default = comma. Choices = comma, tab, whitespace.")

args <- parser$parse_args()
dir_fuma <- remove_trailing_slash(args$dir_fuma)

remove_gene_placeholder <- args$remove_gene_placeholder

posMap_max_distance <- args$posMap_max_distance
n_thread <- args$n_thread

dir_out <- remove_trailing_slash(args$dir_out)
name_out <- args$name_out
delim_out <- args$delim_out

### Function
table.FUMA.positional_mapping <- function(genes_filepath,
                    snps_filepath,
                    posMap_max_distance = 10,
                    n_thread = 1){
    # Require two files
    #   - genes.txt
    #   - snps.txt
    # posMap_max_distance (in Kb): maximum distance between SNP and gene start and end position for positional mapping

    df.snps <- fread(snps_filepath, sep="\t", data.table = F, nThread = n_thread) %>%
        dplyr::select(rsID, chr, pos, posMapFilt)

    df.posMap <- fread(genes_filepath, sep="\t", data.table = F, nThread = n_thread) %>%
        dplyr::select(ensg, symbol, type, chr, start, end, 
                    posMapSNPs, posMapMaxCADD) %>%
        dplyr::rename(`Ensembl ID` = ensg, Gene = symbol, `Gene type` = type,
                    Chr = chr, Start = start, End = end,
                    `N SNP posMap` = posMapSNPs, `Max CADD posMap` = posMapMaxCADD) %>%
        dplyr::mutate(posMap = ifelse(`N SNP posMap` > 0, "Yes", "No"))
                    
    df.posMap_SNP_mapped <- as.data.frame(df.posMap %>%
                                            inner_join(df.snps, by = c("Chr" = "chr"), relationship = "many-to-many") %>%
                                            filter(pos >= Start - posMap_max_distance * 1000 & 
                                                    pos <= End + posMap_max_distance * 1000 &
                                                    posMapFilt == 1) %>%
                                            group_by(`Ensembl ID`, Chr, Start, End) %>%
                                            summarise(`SNP posMap` = paste(rsID, collapse = ";")) %>%
                                            ungroup())
    
    df.posMap <- df.posMap %>%
                    left_join(df.posMap_SNP_mapped, by = c("Ensembl ID", "Chr", "Start", "End"))
    
    return (df.posMap)
}

table.FUMA.eQTL_mapping <- function(genes_filepath,
                    eqtl_filepath,
                    snps_filepath,
                    n_thread = 1){
    # Require two files
    #   - genes.txt
    #   - eqtl.txt
    #   - snps.txt
    
    df.snps <- fread(snps_filepath, sep="\t", data.table = F, nThread = n_thread) %>%
        dplyr::select(uniqID, rsID)

    df.eqtl <- as.data.frame(fread(eqtl_filepath, sep="\t", data.table = F, nThread = n_thread) %>%
            dplyr::select(uniqID, symbol, eqtlMapFilt) %>%
            filter(eqtlMapFilt == 1) %>%
            left_join(df.snps, by = "uniqID") %>%
            dplyr::select(-uniqID))

    df.eqtlMap <- fread(genes_filepath, sep="\t", data.table = F, nThread = n_thread) %>%
        dplyr::select(ensg, symbol, type, chr, start, end,
                    eqtlMapSNPs, eqtlMapminP, eqtlMapminQ, eqtlMapts) %>%
        dplyr::rename(`Ensembl ID` = ensg, Gene = symbol, `Gene type` = type,
                    Chr = chr, Start = start, End = end,
                    `N SNP eqtlMap` = eqtlMapSNPs, `Min P eqtlMap` = eqtlMapminP, 
                    `Tissue eqtlMap` = eqtlMapts) %>%
        dplyr::mutate(eqtlMap = ifelse(`N SNP eqtlMap` > 0, "Yes", "No"))

    df.eqtlMap_SNP_mapped <- as.data.frame(df.eqtlMap %>%
                                            inner_join(df.eqtl, by = c("Gene" = "symbol"), relationship = "many-to-many") %>%
                                            group_by(Gene) %>%
                                            summarise(`SNP eQTL` = paste(rsID, collapse = ";")) %>%
                                            ungroup())
    df.eqtlMap <- df.eqtlMap %>%
        left_join(df.eqtlMap_SNP_mapped, by = "Gene")
    
    return (df.eqtlMap)
}

table.FUMA.chromatin_interaction_mapping <- function(ci_filepath,
                                                    n_thread = 1){
    # Require one file
    #   - ci.txt
    # Lead SNP이 속한 region 1과 interaction이 있는 region 2에 overlap되는 유전자들이 mapping됨.
    # 그래서 하나의 lead SNP이 속한 region 1과 interaction이 있는 region 2와 overlap되는 유전자들이 2개 이상일 수도 있고,
    # 또, 여러 lead SNP이 같은 region 1에 속할 수도 있고,
    # 또, 다른 region1이 같은 region2와 interaction이 있을 수 있음.
    # 그래도 mapping된 unique 유전자에 어떤 lead SNP들, 그리고 tissue/cell에서 mapping 되었는지 정리함.

    df.ciMap <- as.data.frame(fread(ci_filepath, sep="\t", data.table = F, nThread = n_thread) %>%
        dplyr::select(SNPs, genes, `tissue/cell`, ciMapFilt) %>%
        filter(ciMapFilt == 1) %>%
        tidyr::separate_longer_delim(genes, delim = ":") %>%
        distinct(SNPs, genes, .keep_all = T) %>%
        group_by(genes) %>%
        summarise(`SNP ci` = paste0(SNPs, collapse = ";"),
                  `Tissue ci` = paste0(`tissue/cell`, collapse = ";")) %>%
        mutate(ciMap = "Yes") %>%
        dplyr::rename(`Ensembl ID` = genes))
    
    return (df.ciMap)
}
###


file_list <- list.files(dir_fuma,
                    full.names = TRUE)
potential_file_list <- c("genes.txt", "snps.txt",
                        "eqtl.txt",
                        "ci.txt")
potential_file_exists <- potential_file_list %in% file_list


for (filename in potential_file_list){
    print(paste0(filename, " : ", ifelse(file.exists(paste0(dir_fuma, "/", filename)), "exists", "does not exist")))
}

### Start with positional mapping since it is always performed by FUMA
df.pos <- table.FUMA.positional_mapping(genes_filepath = paste0(dir_fuma, "/genes.txt"),
                                    snps_filepath = paste0(dir_fuma, "/snps.txt"),
                                    posMap_max_distance = posMap_max_distance,
                                    n_thread = n_thread)

df_merged <- df.pos

### eQTL mapping
df.eqtl <- data.frame()
if (file.exists(paste0(dir_fuma, "/genes.txt")) &
    file.exists(paste0(dir_fuma, "/eqtl.txt")) &
    file.exists(paste0(dir_fuma, "/snps.txt"))){
    df.eqtl <- table.FUMA.eQTL_mapping(genes_filepath = paste0(dir_fuma, "/genes.txt"),
                                       eqtl_filepath = paste0(dir_fuma, "/eqtl.txt"),
                                       snps_filepath = paste0(dir_fuma, "/snps.txt"),
                                       n_thread = n_thread)
    
    df_merged <- merge(df_merged, df.eqtl, by = c("Ensembl ID", "Gene", "Gene type", "Chr", "Start", "End"), all = T)
}


### Chromatin interaction mapping
df.ci <- data.frame()
if (file.exists(paste0(dir_fuma, "/ci.txt"))){
    df.ci <- table.FUMA.chromatin_interaction_mapping(ci_filepath = paste0(dir_fuma, "/ci.txt"),
                                                      n_thread = n_thread)
    
    df_merged <- merge(df_merged, df.ci, by = "Ensembl ID", all = T) %>%
        replace_na(list(ciMap = "No"))
}

### Add ` to Gene name
if (!remove_gene_placeholder){
    df_merged <- df_merged %>%
        mutate(Gene = paste0("`", Gene))
}


### Save
write.table(df_merged, 
            paste0(dir_out, "/", name_out, ".", delim_extension_map[delim_out]),
            sep=delim_map[delim_out], row.names = F, quote = T)